### PR TITLE
chore: trim string catalogs after updating localizations - WPB-10942

### PR DESCRIPTION
**Pull request by caldrian**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-10942" title="WPB-10942" target="_blank">BugWPB-10942</a>  Fix AppStore Connect warning regarding invalid localisation
  </td></table>
  <br />
 

### Issue

Since we use a string catalog, the crowdin action imports many more localisations than the app is configured for.
When uploading a build to AppStore connect we receive a warning: ITMS-90176: Unrecognized Locale

This PR aims for removing the all but whitelisted localisations before creating a build.

*Changes:*
- Disable creating a PR for the crowdin action, but instead
- run a script to trim the string catalog and if check for changes after the file has been fixed.

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [x] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

**Review by agisilaos — APPROVED** [review]

File: .github/workflows/_reusable_app_release.yml [hunk 1]
```diff
@@ -162,6 +161,8 @@ jobs:
           aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
           aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
           aws-region: eu-west-1
+      - name: Trim unused localizations from string catalogs
+        run: find . -name "*.xcstrings" | xargs xcrun swift run --package-path ./scripts TrimStringCatalogs
```
Comment (netbe) [inline .github/workflows/_reusable_app_release.yml]: so we trim before submitting?
what about trimming when we do update the strings, so during the crowdin job?

File: SourceryPlugin/.swiftpm/xcode/package.xcworkspace/xcshareddata/IDETemplateMacros.plist
Comment (netbe) [inline SourceryPlugin/.swiftpm/xcode/package.xcworkspace/xcshareddata/IDETemplateMacros.plist:1]: why the change here ?

File: scripts/TrimStringCatalogs/main.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+               "localizations" : {
+                 "ar" : {
+                   "stringUnit" : {
+                     "state" : "translated",
+                     "value" : "حفظ في الأرشيف"
+                   },
+                   ...
+         */
+
+        var json = try JSONSerialization.jsonObject(with: data, options: []) as! [String: Any]
```
Comment (netbe) [inline scripts/TrimStringCatalogs/main.swift:75]: really cool to see swift script 😎

**Review by netbe — COMMENTED** [review]
Looks good, but added a question before final go

File: .github/workflows/_reusable_app_release.yml (hunk 1 above)
Comment (caldrian) [reply .github/workflows/_reusable_app_release.yml]: I considered that, but the update action runs on ubuntu which probably (~haven't verified~ [ubuntu-latest)](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md) doesn't have any swift runtime installed.

File: scripts/TrimStringCatalogs/main.swift (hunk 2 above)
Comment (caldrian) [reply scripts/TrimStringCatalogs/main.swift:75]: I agree, it's also nice to have unit tests for scripts.

File: SourceryPlugin/.swiftpm/xcode/package.xcworkspace/xcshareddata/IDETemplateMacros.plist
Comment (caldrian) [reply SourceryPlugin/.swiftpm/xcode/package.xcworkspace/xcshareddata/IDETemplateMacros.plist:1]: The symlinks were wrong.
While in our Xcode projects (like Wire-iOS, WireSyncEngine etc.) the path to the linked file is four levels up, while from our packages it's five levels up.

File: .github/workflows/crowdin.yml [hunk 3]
```diff
@@ -13,17 +13,18 @@ concurrency:
 
 jobs:
   synchronize-with-crowdin:
-    runs-on: ubuntu-latest
+    runs-on: ubuntu-22.04
+    container: swift:5.10.1
```
Comment (caldrian) [inline .github/workflows/crowdin.yml:17]: Pinning the versions so that it doesn't update automatically.

File: .github/workflows/_reusable_app_release.yml (hunk 1 above)
Comment (caldrian) [reply .github/workflows/_reusable_app_release.yml]: I was able to achieve it now.

File: .github/workflows/crowdin.yml [hunk 4]
```diff
@@ -13,31 +13,55 @@ concurrency:
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          git config --global --add safe.directory "$GITHUB_WORKSPACE"
+          find . -name "*.xcstrings" | tee string-catalogs.txt
+          xargs swift run --package-path ./scripts TrimStringCatalogs < string-catalogs.txt
+          xargs git add < string-catalogs.txt
+          # abort if there are no changes
+          if git diff --quiet && git diff --cached --quiet; then echo "No changes to commit."; exit 0; fi
+          git commit --no-verify -m "trim string catalogs"
```
Comment (netbe) [inline .github/workflows/crowdin.yml]:
```suggestion
          git commit --no-verify -m "chore: trim string catalogs"
```

File: .github/workflows/crowdin.yml [hunk 5]
```diff
@@ -13,31 +13,55 @@ concurrency:
...
           token: ${{ secrets.CROWDIN_API_TOKEN }}
           config: crowdin.yml
 
           upload_sources: true
           download_translations: true
           upload_translations: false
-
-          create_pull_request: true
+          create_pull_request: false
+          push_translations: true
```
Comment (netbe) [inline .github/workflows/crowdin.yml:35]: why do we need this ?

Comment (caldrian) [reply .github/workflows/crowdin.yml:35]: The crowdin action populates the string catalog file with many unused translations. Since then it detects a diff it would create a PR.
However, we instead want to remove the unused translates and then check again, if there is still a diff, and only create a PR if needed.

File: .github/workflows/crowdin.yml [hunk 6]
```diff
@@ -13,31 +13,55 @@ concurrency:
...
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Trim unused localizations from string catalogs
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          git config --global --add safe.directory "$GITHUB_WORKSPACE"
+          find . -name "*.xcstrings" | tee string-catalogs.txt
+          xargs swift run --package-path ./scripts TrimStringCatalogs < string-catalogs.txt
+          xargs git add < string-catalogs.txt
```
Comment (netbe) [inline .github/workflows/crowdin.yml:57]: why do we save this file?

File: .github/workflows/crowdin.yml (hunk 5 above)
Comment (netbe) [reply .github/workflows/crowdin.yml:35]: is that a bug at the end of the crowdin action, can we open an issue or fix it there ?

File: .github/workflows/crowdin.yml [hunk 7]
```diff
@@ -13,31 +13,71 @@ concurrency:
...
+          git push --no-verify origin "$HEAD_BRANCH"
+
+          # abort if there no changes to be merged
+          if git diff --quiet "origin/$BASE_BRANCH"; then
+          echo "No changes compared to the base branch, aborting."
+          exit 0
+          fi
+
+          # abort if the pr already exists
+          apt update && apt install -y gh
```
Comment (netbe) [inline .github/workflows/crowdin.yml:76]: should we cache this install?

File: .github/workflows/crowdin.yml (hunk 6 above)
Comment (caldrian) [reply .github/workflows/crowdin.yml:57]: Just to not run the find command twice.
It will be discarded then.

File: .github/workflows/crowdin.yml (hunk 5 above)
Comment (caldrian) [reply .github/workflows/crowdin.yml:35]: We should definitely reach out to them.

File: .github/workflows/crowdin.yml (hunk 7 above)
Comment (caldrian) [reply .github/workflows/crowdin.yml:76]: I can only think of one way: by creating our own docker image. But we would have to update it regularly.

Comment (netbe) [reply .github/workflows/crowdin.yml:76]: I am not sure it's worth the optim, let's leave it for now

**Review by netbe — APPROVED** [review]
Looks good

File: .github/workflows/crowdin.yml (hunk 5 above)
Comment (netbe) [reply .github/workflows/crowdin.yml:35]: can you follow this to their repo with an issue?

Comment (caldrian) [reply .github/workflows/crowdin.yml:35]: https://github.com/crowdin/github-action/issues/245
